### PR TITLE
[BEAM-4018] Add a ByteKeyRangeTracker based on RestrictionTracker for SDF

### PR DESCRIPTION
**Pull request by iemejia**

### CODE REVIEW

**Comment by iemejia** [comment]
Pushed the tracker first to make the review simpler.

**Comment by jkff** [comment]
@chamikaramj Could you do the first round?

**Comment by chamikaramj** [comment]
Sure, will take a look.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+  // Utility methods
+  @VisibleForTesting
+  static ByteKey next(ByteKey key, @Nullable ByteKeyRange range) {
+    // An empty ByteKey may represent a lower or upper bound, we use the range to infer it.
+    if (key.isEmpty() && range != null) {
+      if (range.getStartKey().isEmpty()) {
+        return ByteKey.of(0);
+      } else if (range.getEndKey().isEmpty()) {
+        return ByteKey.EMPTY;
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: When would we hit this condition ? If the range.StartKey() is not empty. We would always be calling next for non-empty keys, no ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    return MoreObjects.toStringHelper(this)
+        .add("range", range)
+        .add("lastClaimedKey", lastClaimedKey)
+        .add("lastAttemptedKey", lastAttemptedKey)
+        .toString();
+  }
+
+  // Utility methods
+  @VisibleForTesting
+  static ByteKey next(ByteKey key, @Nullable ByteKeyRange range) {
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Please add a comment explaining the next() method.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      } else if (range.getEndKey().isEmpty()) {
+        return ByteKey.EMPTY;
+      }
+    }
+    return ByteKey.copyFrom(unsignedCopyAndIncrement(key.getBytes()));
+  }
+
+  // Following methods cloned from org.apache.hadoop.hbase.util.Bytes
+  private static byte[] unsignedCopyAndIncrement(byte[] input) {
+    byte[] copy = copy(input);
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Not fully following this method. Could you add a comment ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
```
[the remark this answers is not in the document]
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Agree, but [ByteKeyRange checks that end will always be after the start](https://github.com/apache/beam/blob/3eacf8b8d7301d1bd4db17874a31caf3b9949526/sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKeyRange.java#L286). I do prefer to remove that restriction from ByteKeyRange or at least loosen it to include equal start and end. Do you agree if I do this? I am not sure this will impact the existing code but we can try.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+   * <p>Must be larger than the last successfully claimed key.
+   *
+   * @return {@code true} if the key was successfully claimed, {@code false} if it is outside the
+   *     current {@link ByteKeyRange} of this tracker (in that case this operation is a no-op).
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
```
[the remark this answers is not in the document]
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: What do you suggest ? I took the message from OffsetRangeTracker only to be consistent between trackers.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 1 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Yes you have a point, I will push a new implementation that does not need the range and deal with this weirdness in a cleaner way.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 2 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 3 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
+        (nextKey.equals(range.getEndKey())) ? next(range.getEndKey(), range) : range.getEndKey();
+    ByteKeyRange res = ByteKeyRange.of(nextKey, rangeEndKey);
+    this.range = ByteKeyRange.of(range.getStartKey(), nextKey);
```
[the remark this answers is not in the document]
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Good remark, but I am not sure if the extra check is needed, I think it is already covered by the tryClaim assertions. If you arrive here the lastClaimedKey must be consistent or it will fail before. Probably that is more clear in the tests.

**Comment by iemejia** [comment]
Ok, Updated with a way simpler version of next and dealing with the 'special' case in other methods, it is worth a second look. Thanks a lot for your review @chamikaramj I think it really improved and is more readable. PTAL.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Or we should just allow checkpoint method to return null if a proper residual cannot be provided (since the range has been fully consumed). If we choose that path we should clarify in the interface. @jkff WDYT ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Optimization: I think comparing to startKey is only necessary if lastAttemptedKey == null

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: * Do you mean "range upper limit" rather than "range upper bundle"?
* Could you give an example where this case can be hit in practice?
* This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand the second sentence here too. What does it mean for the user?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import com.google.common.base.MoreObjects;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+
+/**
+ * A {@link RestrictionTracker} for claiming {@link ByteKey}s in a {@link ByteKeyRange} in a
+ * monotonically increasing fashion. The range is a semi-open bounded interval [startKey, endKey)
+ * where the limits are both represented by ByteKey.EMPTY.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand this sentence - do you mean that ByteKey.EMPTY stands for an empty start key and for an infinite end key?

Would it significantly simplify things if ByteKeyRange had the flag "is the end the infinite key", instead of overloading the meaning of the empty key?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        copy[i] = 0;
+      }
+      byte[] out = new byte[copy.length + 1];
+      out[0] = 1;
+      System.arraycopy(copy, 0, out, 1, copy.length);
+      return out;
+    }
+  }
+
+  private static byte[] copy(byte[] bytes) {
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Use Arrays.copyOf() instead of this method?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 7 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Optimization: I think comparing to startKey is only necessary if lastAttemptedKey == null

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 8 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: * Do you mean "range upper limit" rather than "range upper bundle"?
* Could you give an example where this case can be hit in practice?
* This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 9 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand the second sentence here too. What does it mean for the user?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 10 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand this sentence - do you mean that ByteKey.EMPTY stands for an empty start key and for an infinite end key?

Would it significantly simplify things if ByteKeyRange had the flag "is the end the infinite key", instead of overloading the meaning of the empty key?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 11 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Use Arrays.copyOf() instead of this method?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 7 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Optimization: I think comparing to startKey is only necessary if lastAttemptedKey == null

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 8 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: * Do you mean "range upper limit" rather than "range upper bundle"?
* Could you give an example where this case can be hit in practice?
* This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 9 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand the second sentence here too. What does it mean for the user?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 10 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Don't understand this sentence - do you mean that ByteKey.EMPTY stands for an empty start key and for an infinite end key?

Would it significantly simplify things if ByteKeyRange had the flag "is the end the infinite key", instead of overloading the meaning of the empty key?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 11 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Use Arrays.copyOf() instead of this method?

**Comment by jkff** [comment]
(Sorry for the duplicate comments)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Yeah I'm in favor of loosening the restriction - that seems much less invasive than changing semantics of checkpoint().

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 10 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Yes exactly that. ByteKey is quite tricky it is the lowest and the biggest value depending on it is use in the range. See https://github.com/apache/beam/blob/229e69d241049719fb53adcacab85292d876d7f9/sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKey.java#L38
So I am not overloading anything. I am just following the definition of ByteKey (and also this is consistent with the way HBase (and I suppose Bigtable) partition ranges work.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 9 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Oups that was an error I forgot to delete this. At some moment I had not understood that ByteKey used lexicographic order. I will remove he comment since it is pretty much self documented.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 11 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Good one, didn't know that method, thanks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 8 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: > - Do you mean "range upper limit" rather than "range upper bundle"?

Yes, I will fix, good catch, thanks

> - Could you give an example where this case can be hit in practice?

This only happens because of `ByteKeyRange` validation checkArgument validation, see:
https://github.com/apache/beam/blob/229e69d241049719fb53adcacab85292d876d7f9/sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKeyRange.java#L283-L287

I hit this case while reproducing the unit tests based on `OffsetRangeTracker`, see:
https://github.com/iemejia/beam/blob/d28b9222a4c730ab557091334c555aa8fe89bc67/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java#L89-L100

> - This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

Agree, but remember I am limited by `ByteKeyRange` not allowing degenerate ranges so what do you think, should I change `ByteKeyRange` to allow start == end ?

**Comment by iemejia** [comment]
Changed `ByteKeyRange` to accept degenerate ranges as suggested. Way way nicer without the hack.
This one should be the one, PTAL again @jkff

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any key was successfully claimed");
+    ByteKeyRange res = ByteKeyRange.of(next(lastClaimedKey), range.getEndKey());
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Extract next(..) into a variable?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 7 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: I mean, .compareTo() is not free, and the comparison at line 79 is typically unnecessary - it is only necessary for the very first tryClaim() call; for subsequent calls, if the comparison at line 74 passed, then this one will pass too, by transitivity of comparison.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+   * @return The value of the input incremented by one using byte arithmetic. It treats the input
+   *     byte[] as an unsigned series of bytes, most significant bits first.
+   */
+  private static byte[] unsignedCopyAndIncrement(byte[] input) {
+    byte[] copy = Arrays.copyOf(input, input.length);
+    if (copy == null) {
+      throw new IllegalArgumentException("cannot increment null array");
+    } else {
+      for (int i = copy.length - 1; i >= 0; --i) {
+        if (copy[i] != -1) {
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: The number "-1" doesn't look very good in a function called "unsigned..." - are you sure this is correct? Is there a way to make it more obviously correct?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+  @Test
+  public void testNextByteKey() {
+    assertEquals(next(ByteKey.EMPTY), ByteKey.of(0x00));
+    assertEquals(next(ByteKey.of(0x00)), ByteKey.of(0x01));
+    assertEquals(next(ByteKey.of(0x9f)), ByteKey.of(0xa0));
+    assertEquals(next(ByteKey.of(0xff)), ByteKey.of(0x01, 0x00));
+    assertEquals(next(ByteKey.of(0x10, 0x10)), ByteKey.of(0x10, 0x11));
+    assertEquals(next(ByteKey.of(0x00, 0xff)), ByteKey.of(0x01, 0x00));
+    assertEquals(next(ByteKey.of(0xff, 0xff)), ByteKey.of(0x01, 0x00, 0x00));
+  }
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java]: Are there tests for corner cases like "checkpointing produces an empty residual range" (but doesn't fail), "tryClaim in an empty range doesn't throw" (but returns false)?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 13 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: Well -1 == 0xff so I am pretty sure there should not be an issue, I will change it if you think is more readable.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java (hunk 14 above)
Comment (iemejia) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java]: Yes `testCheckpointClaimedLast` produces an empty range and does not fail. And `testTryClaim` covers the second case.

**Comment by iemejia** [comment]
Thanks for the review @chamikaramj and @jkff. It really helped to improve the quality of the tracker.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java (hunk 13 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java]: My worry here was that due to a widening conversion from byte to int we might end up with a comparison of 0xff against 0xffffffff which would fail.
